graphix/sim/density_matrix.py

```diff
@@ -42,7 +43,14 @@ def __init__(self, data=None, plus_state=True, nqubit=1):
             assert np.isclose(nqubit, int(nqubit))
             nqubit = int(nqubit)
             self.Nqubit = nqubit
-            self.rho = np.asarray(data, dtype=complex).reshape((2**nqubit, 2**nqubit))
+
+            # conversion done above
+            self.rho = data.reshape((2**nqubit, 2**nqubit))
+            np.testing.assert_almost_equal(self.rho.trace(), 1, err_msg="The provided matrix does not have unit trace!")
+            # also add test for hermiticity!!! QuTip test? and modify the DM tests that use not hermitian DMs
+            np.testing.assert_equal(
+                scipy.linalg.ishermitian(self.rho), True, err_msg="The provided matrix is not Hermitian!"
```

Comment:
how about using numpy, for example,
`np.testing.assert_allclose(self.rho, self.rho.getH())`?
scipy is not yet on requirements.txt.

Reply:
I see scipy is uesd for testing too; where necessary, we can install scipy specifically for testing, by adding to https://github.com/TeamGraphix/graphix/blob/master/.github/workflows/ci.yml

I've tried so far to keep requirements.txt as minimum as possible, for faster installation and less chance of version/compatibility issues in the future.

Reply:
> how about using numpy, for example, `np.testing.assert_allclose(self.rho, self.rho.getH())`? scipy is not yet on requirements.txt.

I didn't bother thinking about it first since I believed `scipy`was more efficient but I guess 'numpy' is a good way to go. 
Something that might be interesting for performance upgrade: Qutip mentions that it is possible to do a hermiticity check without compuing the conjugate transpose https://qutip.org/features.html. Might be worth having a look.



> I see scipy is uesd for testing too; where necessary, we can install scipy specifically for testing, by adding to https://github.com/TeamGraphix/graphix/blob/master/.github/workflows/ci.yml
> 
> I've tried so far to keep requirements.txt as minimum as possible, for faster installation and less chance of version/compatibility issues in the future.

Of course. It is just used to generate unitaries but it can be easily replace. Indeed from any random matrix building $U U^\dagger$ will yield a random unitary.

Reply:
The random unitary gen is for testing right? Perhaps the best way might be to require scipy for testing (as above), by adding to ci.yml - what do you think?

Of course, when it becomes necessary, I am always up to the idea of adding to requirements.txt too :) just trying to resist at first to keep the dependency as simple as possible.

Reply:
> The random unitary gen is for testing right? Perhaps the best way might be to require scipy for testing (as above), by adding to ci.yml - what do you think?

Yes, only for testing! I've added it to ci.yml, I hope I did it correctly. But does `pip install -e .[test]` install the additional dependencies specific to the tests?

Reply:
I was imagining adding `pip install scipy` to ci.yml but you had a good point here, it's actually here where we should add scipy:
https://github.com/TeamGraphix/graphix/blob/25e59bc5dfa0b3042348233d7b02626c84d2bf92/setup.py#L39C1-L39C77

Reply:
You sure it's better in the setup.py and not the ci? I guess tests are ran at install. 
Then what does `pip install -e .[test]`do if not installing the dependencies in the tests?